pgjdbc/src/main/java/org/postgresql/util/PGrange.java

```diff
@@ -0,0 +1,317 @@
+   *
+   * @param value the value in the range, not {@code null}
+   * @return the string representation of {@code value}
+   */
+  protected abstract String serializeBound(T value);
+
+  @Override
+  public void setValue(String value) throws SQLException {
+    super.setValue(value);
+    if (value.length() < 2) {
+      throw new PSQLException(
+              GT.tr("Conversion to type {0} failed: {1}.", type, value),
+              PSQLState.DATA_TYPE_MISMATCH);
+    }
+    if ("empty".equals(value)) {
+      this.lowerInclusive = false;
+      this.upperInclusive = false;
+      this.lowerBound = null;
+      this.upperBound = null;
+      return;
+    }
+
+
+    if (value.charAt(0) == '[') {
+      this.lowerInclusive = true;
+    } else if (value.charAt(0) == '(') {
+      this.lowerInclusive = false;
+    } else {
+      throw new PSQLException(
+              GT.tr("Conversion to type {0} failed: {1}.", type, value),
+              PSQLState.DATA_TYPE_MISMATCH);
+    }
+
+    if (value.charAt(value.length() - 1) == ']') {
+      this.upperInclusive = true;
+    } else if (value.charAt(value.length() - 1) == ')') {
+      this.upperInclusive = false;
+    } else {
+      throw new PSQLException(
+              GT.tr("Conversion to type {0} failed: {1}.", type, value),
+              PSQLState.DATA_TYPE_MISMATCH);
+    }
+
+    PGtokenizer t = new PGtokenizer(value.substring(1, value.length() - 1), ',');
+    if (t.getSize() == 0 || t.getSize() > 2) {
+      throw new PSQLException(
+          GT.tr("Conversion to type {0} failed: {1}.", type, value),
+          PSQLState.DATA_TYPE_MISMATCH);
+    }
+
+    try {
+      String token = t.getToken(0);
+      if (token.isEmpty()) {
+        this.lowerBound = null;
+      } else {
+        this.lowerBound = parseToken(token);
+      }
+      if (t.getSize() > 1) {
+        this.upperBound = parseToken(t.getToken(1));
+      } else {
+        this.upperBound = null;
+      }
+    } catch (RuntimeException e) {
+      throw new PSQLException(
+              GT.tr("Conversion to type {0} failed: {1}.", type, value),
+              PSQLState.DATA_TYPE_MISMATCH);
+    }
+  }
+
+  protected abstract T parseToken(String token);
+
+  /**
+   * Returns the range type tag used at the start of the binary representation.
+   *
+   * @return the range type tag
+   */
+  protected int getTag() {
+    if (this.isEmpty()) {
+      return EMPTY;
+    }
+    if (this.isLowerInfinite()) {
+      return this.isUpperInfinite() ? BOTH_OPEN : LOWER_OPEN;
+    } else {
+      return this.isUpperInfinite() ? UPPER_OPEN : CLOSED;
+    }
+  }
+
+  boolean isEmpty() {
+    return nullSafeEquals(this.upperBound, this.lowerBound)
+            && (!this.lowerInclusive || !this.upperInclusive);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj == this) {
+      return true;
+    }
+    if (obj == null) {
+      return false;
+    }
+    if (obj.getClass() != this.getClass()) {
+      return true;
```

Comment:
Is it something that you write by hand or use "generate equals/hashCode" in the IDE?

It is strange that you consider "different class" objects to be equal.

Reply:
That's me being stupid and not writing a unit test for #equals. Will fix.